Remove goog.Uri from KML format

probins opened the pull request:
I think this is ready for review now. As discussed in #5486. Fixes #5486.

marcjansen commented on changelog/upgrade-notes.md:
```diff
@@ -33,6 +33,10 @@ Users compiling their code with the library should note that the following typed
 
+#### KML format now uses URL()
+
+The URL constructor is supported by all modern browsers, but not by older ones, such as IE. To use the KML format in such older browsers, a URL polyfill will have to be loaded before use.
```
Which IE versions will need the polyfill?

probins replied:
all

marcjansen replied:
Ouch. Thanks.

bartvde replied:
I wonder if this will lead to lots of questions / inconvenience? I still think we should consider using https://www.npmjs.com/package/url-parse It should simply work on the latest IE out of the box IMHO?

marcjansen replied:
Exactly my thought.

probins replied:
if you don't use URL(), then I would prefer putting the code in the KML format. It's not a big deal, as it's only appending a string to a base url.

bartvde replied:
Yeah that approach would also work for me.

bartvde replied:
But isn't it also translating a relative url into an absolute url in the case of the KML format?

probins replied:
I'll check that out. The tests are a better fit for URL(), as they're testing all the pieces of the uri, so we can keep the polyfill there.

probins replied:
> But isn't it also translating a relative url into an absolute url in the case of the KML format?

the only logic for which `goog.Uri` is used (in 4 places in KML format, mainly for styles/stylemaps, but also for other links) is:

if string is absolute url (starts with http or https), use that, else append string to node.baseURI

This could be added, to the KML format or perhaps better to `ol.uri`. Unlike the tests, which parse a url into its constituent parts, the KML format doesn't need any real parsing of the url other than determining if it's absolute or relative.

I don't think using an npm package is a good idea, as it penalises the majority using modern browsers (extra code overhead) just to support the minority using old browsers.

ahocevar commented:
Ok, to move this forward: I agree with @probins. It makes sense to use the `URL()` constructor and a polyfill in the tests, while adding custom code to a `ol.uri` namespace to support the one use case in `ol.format.KML`.

Sorry for the back and forth on this. @probins, would you still be willing to update this pull request as outlined above, so everyone is happy?

probins commented:
yes, ok. probably tomorrow

probins commented:
ok. Changed so URL() no longer needed in src but still used in tests. I've simplified the code a bit and made it a bit more consistent. `node.baseURI` is always absolute so there's no need for any check on that. So I don't see the need for any function in `ol.uri`.

bartvde commented:
@probins have you checked Firefox? See: https://github.com/openlayers/ol3/issues/3439

probins commented:
don't understand https://github.com/openlayers/ol3/issues/3439#issuecomment-89537262 - `node.baseURI` is not an empty string in Chrome (just double-checked using http://openlayers.org/en/master/examples/kml.html?mode=raw). In fact, I'm not sure the check for `node.baseURI` is actually needed; according to https://developer.mozilla.org/en-US/docs/Web/API/Node/baseURI it should be always set to the absolute uri.

bartvde commented:
hmm maybe this was changed in Chrome then

probins commented:
https://dom.spec.whatwg.org/#dom-node-baseuri - "The baseURI attribute’s getter must return node document’s document base URL, serialized"

bartvde commented:
Code LGTM @probins have you verified that KML unit tests pass in all major browsers that we support?

probins commented:
no, I don't have access to Windows/Mac. If that's the objective, perhaps getting Travis to run tests on Sauce Labs would be a good idea? (Though the advantage of standards-based browsers _should_ be that they should all Just Work. :-))

bartvde commented:
@probins I have 2 test failures in Firefox on your branch:

![screen shot 2016-06-30 at 13 05 32](https://cloud.githubusercontent.com/assets/319678/16486200/594e7e36-3ec3-11e6-9c29-dfec43c1c8d5.png)

probins commented:
yes, well, there is rather a basic error in the code; I have no idea how this passed the tests on Travis. Carrying out more extensive tests before pushing a correction.

probins commented:
@bartvde could you try again?

bartvde commented:
One failure left @probins 

![screen shot 2016-06-30 at 15 21 53](https://cloud.githubusercontent.com/assets/319678/16489560/69201154-3ed6-11e6-978a-d5c5a884da36.png)

probins commented:
ok, I think I now understand your previous comment on `node.baseURI` being null on Chrome. According to the KML spec, "Relative URLs can be used in this tag and are evaluated relative to the enclosing KML file." So if for example one of the files used in OL's KML examples, say http://openlayers.org/en/master/examples/data/kml/2012-02-10.kml, had a NetworkLink `node.BaseURI` would = that filename. However, in the case of this test, the KML file is created from a string, so there is no 'enclosing KML file'. In this case in Chrome `node.baseURI` is indeed null. It looks like Firefox is using `document.baseURI` instead.

Not sure this test is meaningful. If you want to test relative URLs, ISTM it would be better to lead the test from a file rather than from a string.

bartvde commented:
Would it make more sense then to change to test to be more meaningful? If you think that is the way forward, I agree.

probins commented:
> If you want to test relative URLs

it's not just the relative ones: if `node.baseURI` is null, the code skips the relative/absolute check and just outputs the incoming value. This would also explain why Travis passed my previous code even though it had an error in it: it was just testing that a==a.

If we change the test, I think that should be in a separate PR.

probins commented:
I've been checking this URI subject in more detail. To summarise:
* KML format is the only place in the library where `goog.Uri` is used
* there are 2 types of URI supported in the KML format:
  * `<styleUrl>` elements refer to `<Style>` or `<StyleMap>` elements. These are prefixed with '#', and can be in the same file, or in an external file where the reference can be an absolute or relative Url: examples from KML spec `<styleUrl>#myIconStyleID</styleUrl>`, `<styleUrl>http://someserver.com/somestylefile.xml#restaurant</styleUrl>`, `<styleUrl>eateries.kml#my-lunch-spot</styleUrl>`
  * `<href>` elements occur in Icon and Link elements. These are standard links to an external file with an absolute or relative Url
* these are handled by `readStyleUrl_`, `readSharedStyle_`, `readSharedStyleMap_`, `readURI_`
* there should be tests for all these cases, but only StyleUrl in the same file, absolute Icon, and absolute and relative Link hrefs are currently being tested
* the code uses `node.baseURI`; the normal case is that the KML is in a file, and in this case `node.baseURI` will be the absolute Url of that file. However, the KML could also come from a string, for example, when created programatically, and in this case browsers seem to vary; in Chrome it's null, but Firefox seems to set it to `document.baseURI`. If it's null, the code does not use `goog.Uri` to create an absolute url, but simply returns what was passed in
* the tests create KML from a string; PhantomJS uses QtWebkit, which like Chrome sets `node.baseURI` to null. So although it looks like the tests are currently passing, in fact they're simply testing that what was passed in gets returned, which is always the case; this isn't very useful

I don't think this PR is currently handling all the above cases correctly, so I'll review this and resubmit.

There is though the question of what baseURI should be used if the KML comes from a string. Should it follow Firefox and always use `document.baseURI`?

Although this doesn't have anything to do with replacing `goog.Uri`, tests should be added to fetch styles in external files. As the common case is that the KML is in a file, they should test this. There should also be tests for external links in KML read from a file. Should these replace the current KML from strings, or be in addition?

probins commented:
what I'll do is separate the 2 parts of this PR:

1. this PR can be for replacing `goog.Uri` in kml; although adding new tests isn't strictly for that, we need proper tests to be sure the new code works properly
2. I'll create a new PR for replacing `goog.Uri` in the non-kml tests

probins commented:
looks like there's another issue in kml url handling: `readStyleUrl_()` resolves the url relative to `node.baseURI`, but this is only executed when the `styleUrl` element is a child of a `StyleMap`, not when it is a child of a `Placemark`.

ahocevar commented:
I still agree with this solution. Thanks @probins, and sorry for the back and forth. @marcjansen, would you agree too if we had a prominent place in the documentation that lists supported browsers and required polyfills?

probins commented:
bear in mind too that use of these old browsers is slowly but steadily decreasing. According to http://gs.statcounter.com/#all-browser-ww-monthly-201506-201606 IE is now below 10%.

I'm currently looking into adding tests. This can either be added to this PR or put in a separate one.

ahocevar commented:
@probins, please proceed with the tests as you prefer. I guess the only thing really required to finalise this pull request is to update the polyfills list in `README.md` and `docs/quickstart.hbs`.

probins commented:
docs updated. It will probably be next week before I can get around to the tests. In the meantime, there is #5558 to review.

probins commented:
ready for review again. I added a couple of tests, but it looks like `node.baseURI` is null when reading in files too, so I'm not sure the NetworkLink test I've added is worth the bother; this may well be just a limitation of the test environment. I also found that, although `node.baseURI` is set when using `XMLHttpRequest()`, it is null when you `fetch()` a file, at least in Chrome - this could well be a little bug in Chrome's implementation of `fetch()`. I have tested in a proper web environment and apart from this it all seems to work correctly.

ahocevar commented:
@probins Not sure I entirely understand the implications of your comment above. To me, it appears that everything works as expected now. Are you saying that in Chrome, resolving relative network links will fail when the KML file was retrieved with `fetch()`? Also, what limitation of the test environment are you suspecting?

probins commented:
> Are you saying that in Chrome, resolving relative network links will fail

if `node.baseURI` is null, then the `URL()` (or `goog.Uri`) logic is skipped, which means that relative URLs (for styles and icons as well as network links) are resolved relative to the document rather than the kml file, which may not be the same. I thought at first that this was only in the case where the kml is created from a string, but there seems to be a wider problem. I've not tested in other browsers to see what they do. In practice, this may not matter too much, as I suspect that few links in kml files will use a relative url anyway.

ahocevar commented:
Thanks a lot @probins. The code looks great, and I'm just going to merge it.

In case issues come up, they can be handled in separate pull requests, and the respective fixes should include tests.
